### PR TITLE
hw: mu: Add mu-write implementation

### PR DESCRIPTION
**Pull request by crazoes**
Add implementation for MU write in order to
communicate with the host.

Signed-off-by: Diana Cretu <dianacretu2806@gmail.com>
Signed-off-by: Shreeya Patel <shreeya.patel23498@gmail.com>

### CODE REVIEW

**Comment by crazoes** [comment]
This extends the work that was being done here earlier - https://github.com/thesofproject/qemu/pull/40

**Comment by dbaluta** [comment]
@crazoes please squash this PR with #40. Thus you will have one big commit with patches squashed. The explanation in this commit is very good, you can keep it at the end of the commit for the squashed commits.

By squash I mean this: https://www.devroom.io/2011/07/05/git-squash-your-latests-commits-into-one/

**Comment by crazoes** [comment]
@dbaluta it is already squashed with that PR's commits

**Comment by dbaluta** [comment]
@crazoes Indeed, you are right! thanks!

File: include/hw/adsp/mu.h [hunk 1]
```diff
@@ -31,4 +31,5 @@
 /* General Purpose Interrupt Request */
 #define IMX_MU_xCR_GIRn(x)      BIT(16 + (3 - (x)))
 
+
```
Comment (dbaluta) [inline include/hw/adsp/mu.h]: No need for this new line here.

File: hw/adsp/dsp/imx8-mu.c [hunk 2]
```diff
@@ -51,12 +51,69 @@ static void mu_write(void *opaque, hwaddr addr,
     struct adsp_io_info *info = opaque;
     struct adsp_dev *adsp = info->adsp;
     struct adsp_reg_space *space = info->space;
+    struct qemu_io_msg_irq irq;
+    uint64_t aux;
 
     log_write(adsp->log, space, addr, val, size,
```
Comment (dbaluta) [inline hw/adsp/dsp/imx8-mu.c:57]: One more thing. Aux is not a really good name for a variable in this case.

use xsr and xcr for reading CR and SR registers.

File: hw/adsp/dsp/imx8-mu.c [hunk 3]
```diff
@@ -51,12 +51,69 @@ static void mu_write(void *opaque, hwaddr addr,
...
+    default:
+        break;
+    }
+}
+
+void adsp_imx8_irq_msg(struct adsp_dev *adsp, struct qemu_io_msg *msg)
+{
+    struct adsp_io_info *info = adsp->mu_b;
+    uint64_t xcr;
+
```
Comment (dbaluta) [inline hw/adsp/dsp/imx8-mu.c:102]: Here you should define two variables xcr and xsr. I know that it works with one but the code is confusing.

File: hw/adsp/dsp/imx8-mu.c [hunk 4]
```diff
@@ -51,12 +51,69 @@ static void mu_write(void *opaque, hwaddr addr,
...
+    xcr = info->region[IMX_MU_xCR >> 2];
+
+    /* reply arrived from host */
+    if (xcr & IMX_MU_xCR_GIRn(1)) {
+        /* set pending bit for reply on DSP */
+        xcr = info->region[IMX_MU_xSR >> 2] & ~0;
+        xcr |= IMX_MU_xSR_GIPn(1);
```
Comment (dbaluta) [inline hw/adsp/dsp/imx8-mu.c]: Use xsr here.

File: hw/adsp/dsp/imx8-mu.c [hunk 5]
```diff
@@ -51,12 +51,69 @@ static void mu_write(void *opaque, hwaddr addr,
...
+        info->region[IMX_MU_xSR >> 2] = xcr;
+    } else {
+        /* new message arrived from host, so activate pending bit */
+        xcr = info->region[IMX_MU_xSR >> 2] & ~0;
+        xcr |= IMX_MU_xSR_GIPn(0);
+        info->region[IMX_MU_xSR >> 2] = xcr;
```
Comment (dbaluta) [inline hw/adsp/dsp/imx8-mu.c]: Use xsr here.

**Review by dbaluta — APPROVED** [review]

**Comment by dbaluta** [comment]
LGTM, @lgirdwood this is good to merge.

**Comment by marc-hb** [comment]
For some reason this is not compatible with Ubuntu 20.04? See https://github.com/thesofproject/sof/issues/4186

**Comment by fredoh9** [comment]
This is not working either Ubuntu 18.04 or 20.04. We haven't made sofqemu image for a year. This is first time image update after this commit.

**Comment by marc-hb** [comment]
Thanks @fredoh9 , sorry for the confusion.

So this commit is the last commit in `qemu:master`. @fredoh9 how long would it take to upgrade sofqemu to Ubuntu 20.04 _and_ `qemu:master~1`?

**Comment by fredoh9** [comment]
Building sofeqmu is faster then sof image. In my build machine, it takes about 10 min.

**Comment by marc-hb** [comment]
Hi @fred, I was asking how much effort it would be for you, sorry for the confusion.

**Pull request merged**